Add String.hash and String.seeded_hash

ctk21 opened the pull request:
This PR adds a specialization of `caml_hash` for type `string` and adds it to the stdlib `String` module. 

There are two motivations for this change:
 (i) It allows the user to easily improve the performance of their string hash tables by using the new string hash function.
 (ii) It allows the user to easily create a string based hash table with less boiler plate `Hashtbl.Make(String)`. Within stdlib you can already do `Map.Make(String)` and when using zarith `Hashtbl.Make(Z)` works.

We can add further definitions of `hash` across stdlib to get (ii) for more types, which could be desirable.

The performance impact of the change will vary depending on the length of the string being hashed. Micro-benchmarking the change, I see the following results:
```
string length    old (s)  new (s)  improvement (%)
            8      0.225    0.156             30.6
           16      0.252    0.194             22.9
           32      0.322    0.253             21.3
           64      0.461    0.393             14.8
          128      0.775     0.74             4.52
          256       1.45     1.36             6.02
          512        2.8     2.73             2.55
```
with this code:
```ocaml
let rec test fn test_str n_iter acc =
  match n_iter with
    | 0 -> acc
    | n ->
      let h = fn test_str in
      test fn test_str (n_iter-1) (acc+h)
```
and 10,000,000 iterations on a x86 E5-2430L v2.

These results suggest a reasonable performance improvement for strings up to 128 characters in length. These results are consistent with (the identical) specialization used in the Base library:
https://github.com/janestreet/base/blob/master/src/string.ml#L839

I also tried the change (from `Hashtbl.hash` to `String.hash`) in the F* compiler and it showed a 4-5% improvement in the F* user library benchmark (in aggregate across the whole suite).

ctk21 committed:
caml_hash_string specialization for String.hash

cfcs commented on stdlib/string.mli:
```diff
@@ -316,6 +316,18 @@ val equal: t -> t -> bool
+    {!Hashtbl.hash}. Along with the type [t], this function [hash]
+    allows the module [String] to be passed as argument to the functor
+    {!HashTbl.Make}.
+    @since 4.10.0 *)
```
This will leave people who follow the (now recommended?) pattern of `Hashtbl.Make(String)` vulnerable to the hash table collision attacks that the seeded hashing was intended to thwart.
Would it make sense to have a note about this in the docstring, explaining that if untrusted input is used as keys when inserting, a random seed should be used?
See:
- the [docstring for Hashtbl.create](https://caml.inria.fr/pub/docs/manual-ocaml/libref/Hashtbl.html)
- [our initial patch re: this problem](https://github.com/ocaml/ocaml/issues/5572)
- more discussion in the [PR for SipHash](https://github.com/ocaml/ocaml/pull/24)

alainfrisch replied:
It's unfortunate that Hashtbl.SeededHashedType expects a `val hash: int -> t -> int` and not `val seeded_hash: int -> t -> int`.  This prevents providing modules which can be passed to both `Hashtbl.Make` and `Hashtbl.MakeSeeded`.

alainfrisch replied:
If the future is to deprecate `Hashtbl.Make` in favor of `Hashtbl.MakeSeeded` (which still allows creating non-randomized tables), one should expose only `val String.hash: int -> t  -> int` (and perhaps revisit Uchar.hash and Float.hash accordingly).

dra27 replied:
I agree it's better that all the tests and examples use `Hashtbl.MakeSeeded`, yes - `Hashtbl.Make` should only be used when one fully understands the consequences (which, for the F* case this comes from, they do!). Documentation changes, if any, probably want to go on the `Hashtbl.Make` functor itself, though?

cfcs replied:
I agree re: `seeded_hash` vs `hash`, but it seems a little bit out of scope for this PR to rectify this problem outside of the `String` module.

Re: documentation changes: IIRC the `Hashtbl.Make` functor does not have any documentation (the documentation resides in the outer module or in the `S` signature)?
I think if we can refer to `Hashtbl.Make` here we can also refer to the documentation about the potential problems related to using that functor (with unseeded hashes).

cfcs commented on testsuite/tests/lib-string/test_string.ml:
```diff
@@ -40,6 +40,11 @@ let () =
+
+let () =
+  assert((Hashtbl.hash raw_string) = (String.hash raw_string))
+;;
```
Should there be (or is there) a test that makes sure that the hashing function doesn't just return the same (e.g. 0) for all values?

alainfrisch commented on stdlib/string.ml:
```diff
@@ -202,6 +202,10 @@ type t = string
 let compare (x: t) (y: t) = Stdlib.compare x y
 external equal : string -> string -> bool = "caml_string_equal" [@@noalloc]
 
+external seeded_hash_param : int -> 'a -> int = "caml_hash_string" [@@noalloc]
```
`caml_string_hash` would be coherent with `caml_string_equal`

alainfrisch commented on stdlib/string.mli:
```diff
@@ -316,6 +316,18 @@ val equal: t -> t -> bool
 
+val hash : t -> int
+(** The hash function for strings, with the same specification as
+    {!Hashtbl.hash}. Along with the type [t], this function [hash]
```
- I'd say "A hash function", not "The hash function", unless you mean that that function returns the same result as `Stdlib.hash` (but then it should be made more explicit).

- Along with the type [t] and the value [equal].  But I'm not sure mentioning all that is useful (there is no such info for Uchar.hash or Float.hash).

alainfrisch commented on stdlib/string.mli:
```diff
@@ -316,6 +316,18 @@ val equal: t -> t -> bool
+(** The hash function for strings, with the same specification as
+    {!Hashtbl.hash}. Along with the type [t], this function [hash]
+    allows the module [String] to be passed as argument to the functor
+    {!HashTbl.Make}.
```
```suggestion
    {!Hashtbl.Make}.
```

ctk21 committed:
Address review comments: caml_string_hash rename, changes to use HashTbl.MakeSeeded functors the as the default, explictly pin string hash test output, doc updates, implementation for StringLabels

ctk21 commented:
Thank you all for the review suggestions. I'm assuming that the comments are broadly in favour of the change. 

The consensus seems to be to support `Hashtbl.MakeSeeded` going forward. I have:
 - removed reference to `Hashtbl.Make` in favour of `Hashtbl.MakeSeeded` in the documentation
 - moved to `unseeded_hash`/`hash` from `hash`/`seeded_hash` to allow support with `Hashtbl.MakeSeeded`
 - updated the `htbl.ml` tests to use `Hashtbl.MakeSeeded`
 - fixed up the `StringLabels` mli to match
 - moved from `caml_hash_string` to `caml_string_hash` to match `caml_string_equal`
 - updated the string tests to be explicit about their output

Let me know if there is anything else that is needed.

alainfrisch commented:
Sorry for the long delay, and yes, I'm in favor of the accepting the change (more than just "broadly in favor", actually). My only concern is about supporting Hashtbl.Make vs Hashtbl.MakeSeeded. As often when discussing Stdlib extension, we need to decide between maintaining coherence with existing modules (`Uchar` and `Float` being designed to be passed to `Hashtbl.Make`) or to go with the latest recommendations (always use `MakeSeeded`)

I'd like to hear about others' opinion (in particular @dbuenzli 's).  We could:

  1. Accept incoherence with Uchar/Float.
  2. Plan to switch `Uchar.hash/Float.hash` to the seeded version  -- breaking change, but perhaps it's still doable.
  3. Change `Hashtbl.SeededHashedType` to use `seeded_hash` instead of `hash` (thus allowing modules to be passed to both `Hashtbl` and `SeededHash`) -- probably breaks too much code.
  4. Deprecate `Hashtbl.MakeSeeded` and create a new variant of it which uses a `seeded_hash` function as input.
  5. Revert to this PR to its original state (unseeded version).

I think 1 is ok for now, but I'd like to discuss whether we'd go rather for 2 or 4 (3 seems unrealistic) or just stick with the incoherence.

dbuenzli commented:
Regarding the topic of hashes itself, I'm not the best person to ask about it, I almost never use them. I added `Uchar.hash` on @bobot's [request](https://github.com/ocaml/ocaml/pull/80#discussion_r14857426) (also we willingly left them out of the `Option` and `Result` modules because one needs a strategy for that, see @aalekseyev's comment [here](https://github.com/ocaml/ocaml/pull/1940#issuecomment-409668832)).

Regarding the more general stdlib strategy, I'm in favour of a language that entices you to do the right things from a security perspective. As such I'd rather do 2. and have people in the broader eco-system only implement `SeededHashedType` by adopting the following strategy: 

1. Deprecate `HashedType` and `Hashtbl.Make`.
1. Make all stdlib modules that do implement `HashedType` implement `SeededHashedType` (breaking change).
1. Add the functor `module MakeUnseeded : functor (H : SeededHashedType) -> S with type key = H.t` for those people which have a legitimate reason to have non randomized hash table.

alainfrisch commented:
3 might not be needed, it's possible to avoid the random seed even with MakeSeeded (`create ~random:false`).

dbuenzli commented:
Re 3. I think I rather have a module that guarantees me to always give me unrandomized hashtables than having it to specify at each creation site. It's also an easy first target for audits before you go on check all these `MakeSeeded(H).create` do not use `~random:false`.

dbuenzli commented on stdlib/stringLabels.mli:
```diff
@@ -273,6 +273,17 @@ val equal: t -> t -> bool
+(** A seeded hash function for strings, with the same output value as
+    {!Hashtbl.seeded_hash}. This function allows the module [StringLabels]
+    to be passed as argument to the functor {!Hashtbl.MakeSeeded}.
+    @since 4.10.0 *)
```
I don't think you should put a label here if you want to be able to pass it to `Hashtbl.MakeSeeded`. Maybe if you wanted to pass it to 
`MoreLabels.Hashtbl.MakeSeeded` but no, since in that one  `module type SeededHashedType = Hashtbl.SeededHashedType`. 

I think this shows once more why the whole labelled `stdlib` is not really sunstainable design wise and should rather [die](https://github.com/ocaml/ocaml/issues/7400).

dbuenzli commented on stdlib/string.mli:
```diff
@@ -316,6 +316,17 @@ val equal: t -> t -> bool
 (** The equal function for strings.
     @since 4.03.0 *)
 
+val unseeded_hash : t -> int
```
I think this function should be removed.

dbuenzli replied:
Just to be clear I think we should provide `Hashtbl.no_seed` and people should use `M.hash Hashtbl.no_seed` with the seeded version to get this.

dbuenzli commented:
Re 3 again it's also a nicer upgrade path for these `Hashtbl.Make` users.

alainfrisch commented:
Thanks @dbuenzli for all comments.  I haven't been exposed to the seeded variants and always use the non-seeded functor.  Having some inputs from people who feel more involved with the Seeded stuff could be useful as well.  Perhaps @xavierleroy ?

xavierleroy commented:
I think that seeded hashes are good and their use should be encouraged.  (Remember: the lack of seeding in OCaml's hash tables got us slapped with a CVE, and rightly so I think.)

In retrospect, I made a mistake when defining the `Hashtbl.SeededHashedType`: the hash function should have been called `seeded_hash`, not `hash`, so that a single module can implement both signatures `SeededHashedType` and `HashedType`.  Unfortunately, that mistake was not spotted during review.

In terms of the choices offered by @alainfrisch I would be in favor of 3 (if an OPAM grep convinces us that `Hashtbl.MakeSeeded` is used so rarely that we can afford a breaking change) or 4 (otherwise). 

In the meantime, I think the present PR should be suspended, and the `hash` functions already present in `Float` and `Uchar` be viewed as a dubious feature that should be revised later (by adding seeded hash variants).

dbuenzli commented:
@xavierleroy if you want to encourage the use of seeded hashes, I think it's better to require from users to implement only the seeded hash and define a functor using it with a fixed seed to provide an unseeded hashtable rather than letting them choose between implementing `hash` or `seeded_hash` and/or bother them with defining both.

xavierleroy commented:
> if you want to encourage the use of seeded hashes, I think it's better to require from users to implement only the seeded hash

I hear that.  But it feels like a bigger, less backward-compatible change.  Am I wrong?

dbuenzli commented:
Not really basically we need to provide `MakeUnseeded : functor (H : SeededHashedType) -> S` and people can then just implement `SeededHashedType`. 

A second step would be to eventually deprecate `Make (H : SeededType)` and `SeededType` to make a stronger signal but we can do without if you don't like it.

alainfrisch commented:
> In the meantime, I think the present PR should be suspended

I tend to agree.  But what about adding actual automatic specialization of the generic hash primitive(s) on strings, similar to what is done for the generic comparisons?  (This would require adding a %-primitive.)  At least this would allow writing `Hashtbl.Make(struct type t = string let equal (x : t) y = x = y let hash (x: t) = Hashtbl.hash x end)` (or the seeded equivalent) and get the performance boost from the PR description, without having to change the `Stdlib`.

nojb committed:
Tweak

nojb committed:
Changes

nojb reviewed (approved):
LGTM

nojb commented:
Hello @ctk21! Following merge of #11157 this PR can be brought back to life. I took the liberty of rebasing it on top of `trunk`, adding a Changes entry, removing the useless bootstrap, etc. I believe it is ready for merging, please speak up if you have any objections. @alainfrisch: you were involved in the review process, do you mind giving an official approval?

alainfrisch reviewed (approved):
LGTM as well.  Thanks to everyone involved in the discussion!
